Scaling down > 20 instances at once in AWS ASG fails

Hello! 👋 

We are using the `nomad-autoscaler` to effectively scale up for running large batch jobs and then back down again at the end. However we are seeing the autoscaler throw errors when trying to scale down more than 20 instances at a time (anything less works a-ok!)

**Steps To Reproduce:**

1. Create an AWS ASG with a maximum capacity of > 20.
2. Create a policy like this:

```hcl
scaling "my_amazing_scaling" {
  enabled = true
  min     = 0
  max     = 21

  policy {
    cooldown            = "2m"
    evaluation_interval = "30s"

    check "desired" {
      source = "prometheus"
      query  = "21" # <- change this to 0 and see what happens
      strategy "target-value" {
        target = 0
      }
    }

    target "aws-asg" {
      dry-run             = "false"
      aws_asg_name        = "your_asg"
      node_class          = "your_asg"
      node_drain_deadline = "5m"
    }
  }
}
```
3. Wait for boxes to scale up.
4. Change `query` from `"21"` to `"0"`.

**Expected Behaviour:**

The autoscaler should correctly terminate all 21 instances.

**Actual Behaviour:**

The autoscaler fails to terminate the instances with the following error:

```
2021-03-03T18:09:08.148Z [ERROR] policy_eval.worker: failed to evaluate policy: eval_id=<id> eval_token=<id> id=<id> policy_id=<id> queue=cluster err="failed to scale target: failed to perform scaling action: failed to scale in AWS AutoScaling Group: failed to detach intances from Autoscaling Group: ValidationError: List exceeds the maximum number of instances to detach. Please retry your request with fewer instances.
	status code: 400, request id: <id>"
```

This is because there is a maximum limit of 20 instances in a single detch call in the AWS API:

> You can use AttachInstances, DetachInstances, EnterStandby, and ExitStandby with at most 20 instance IDs at a time.

https://docs.aws.amazon.com/autoscaling/ec2/userguide/as-account-limits.html

--- a/plugins/builtin/target/aws-asg/plugin/aws.go
+++ b/plugins/builtin/target/aws-asg/plugin/aws.go
@@ -2,13 +2,14 @@ package plugin
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"time"
 
 	"github.com/aws/aws-sdk-go-v2/aws"
 	"github.com/aws/aws-sdk-go-v2/aws/external"
 	"github.com/aws/aws-sdk-go-v2/service/autoscaling"
-	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	"github.com/hashicorp/nomad-autoscaler/sdk/helper/scaleutils"
 	"github.com/hashicorp/nomad/api"
 )
 
@@ -55,8 +56,7 @@ func (t *TargetPlugin) setupAWSClients(config map[string]string) error {
 		cfg.Credentials = aws.NewStaticCredentialsProvider(keyID, secretKey, session)
 	}
 
-	// Set up our AWS clients.
-	t.ec2 = ec2.New(cfg)
+	// Set up our AWS client.
 	t.asg = autoscaling.New(cfg)
 
 	return nil
@@ -93,18 +93,36 @@ func (t *TargetPlugin) scaleOut(ctx context.Context, asg *autoscaling.AutoScalin
 
 func (t *TargetPlugin) scaleIn(ctx context.Context, asg *autoscaling.AutoScalingGroup, num int64, config map[string]string) error {
 
-	ids, err := t.clusterUtils.RunPreScaleInTasks(ctx, config, int(num))
+	// The AWS plugin utilises the individual scaleutils calls due to the way
+	// in which the TerminateInstanceInAutoScalingGroupRequest call works. We
+	// need to ensure the nodes are part of the target ASG before performing
+	// actions, and draining the nodes just to get an error at this phase is
+	// wasteful and means we need reconciliation.
+	if t.clusterUtils.ClusterNodeIDLookupFunc == nil {
+		return errors.New("required ClusterNodeIDLookupFunc not set")
+	}
+
+	nodes, err := t.clusterUtils.IdentifyScaleInNodes(config, int(num))
 	if err != nil {
-		return fmt.Errorf("failed to perform pre-scale Nomad scale in tasks: %v", err)
+		return err
 	}
 
-	// Grab the instanceIDs once as it is used multiple times throughout the
-	// scale in event.
-	var instanceIDs []string
-
-	for _, node := range ids {
-		instanceIDs = append(instanceIDs, node.RemoteResourceID)
+	nodeResourceIDs, err := t.clusterUtils.IdentifyScaleInRemoteIDs(nodes)
+	if err != nil {
+		return err
 	}
+
+	// Any error received here indicates misconfiguration between the ASG and
+	// the Nomad node pool.
+	instanceIDs, err := instancesBelongToASG(asg, nodeResourceIDs)
+	if err != nil {
+		return err
+	}
+
+	if err := t.clusterUtils.DrainNodes(ctx, config, nodeResourceIDs); err != nil {
+		return err
+	}
+	t.logger.Info("pre scale-in tasks now complete")
 
 	// Create the event writer and write that the drain event has been
 	// completed which is part of the RunPreScaleInTasks() function.
@@ -113,85 +131,130 @@ func (t *TargetPlugin) scaleIn(ctx context.Context, asg *autoscaling.AutoScaling
 
 	// Create a logger for this action to pre-populate useful information we
 	// would like on all log lines.
-	log := t.logger.With("action", "scale_in", "asg_name", *asg.AutoScalingGroupName,
-		"instances", instanceIDs)
+	log := t.logger.With("action", "scale_in", "asg_name", *asg.AutoScalingGroupName)
 
-	// Detach the desired instances.
-	log.Debug("detaching instances from AutoScaling Group")
+	// Run the termination and log the results.
+	result := t.terminateInstancesInASG(ctx, nodeResourceIDs)
+	result.logResults(log)
 
-	if err := t.detachInstances(ctx, asg.AutoScalingGroupName, instanceIDs); err != nil {
-		return fmt.Errorf("failed to scale in AWS AutoScaling Group: %v", err)
-	}
-	log.Info("successfully detached instances from AutoScaling Group")
-	eWriter.write(ctx, scalingEventDetach)
-
-	// Terminate the detached instances.
-	log.Debug("terminating EC2 instances")
-
-	if err := t.terminateInstances(ctx, instanceIDs); err != nil {
-		return fmt.Errorf("failed to scale in AWS AutoScaling Group: %v", err)
-	}
-	log.Info("successfully terminated EC2 instances")
-	eWriter.write(ctx, scalingEventTerminate)
-
-	// Run any post scale in tasks that are desired.
-	if err := t.clusterUtils.RunPostScaleInTasks(ctx, config, ids); err != nil {
-		return fmt.Errorf("failed to perform post-scale Nomad scale in tasks: %v", err)
+	// If we have any failures, perform our revert so we don't leave nodes in
+	// an undesired state.
+	if result.lenFailure() > 0 {
+		t.clusterUtils.RunPostScaleInTasksOnFailure(result.failedIDs())
 	}
 
-	return nil
+	// If we had successfully termination from the ASG, track these activities
+	// until completion. A failure here should not fail the scaling activity as
+	// AWS should honour the contract, it could be a case of there being
+	// slowness in the AWS system and us timing out.
+	if result.lenSuccess() > 0 {
+
+		t.logger.Debug("ensuring AWS ASG activities complete")
+
+		if err := t.ensureActivitiesComplete(ctx, *asg.AutoScalingGroupName, result.activityIDs()); err != nil {
+			log.Error("failed to ensure all activities completed", "error", err)
+		} else {
+			t.logger.Debug("confirmed AWS ASG activities completed")
+		}
+		eWriter.write(ctx, scalingEventTerminate)
+
+		// Run any post scale in tasks that are desired.
+		if err := t.clusterUtils.RunPostScaleInTasks(ctx, config, result.successfulIDs()); err != nil {
+			return fmt.Errorf("failed to perform post-scale Nomad scale in tasks: %v", err)
+		}
+	}
+
+	if result.lenFailure() > 0 && result.lenSuccess() > 0 {
+		log.Warn("partial scaling success",
+			"success_num", result.lenSuccess(), "failed_num", result.lenFailure())
+		return nil
+	}
+	return result.errorOrNil()
 }
 
-func (t *TargetPlugin) detachInstances(ctx context.Context, asgName *string, instanceIDs []string) error {
+// instancesBelongToASG checks that all the instances identified for scaling in
+// belong to the target ASG.
+func instancesBelongToASG(asg *autoscaling.AutoScalingGroup, ids []scaleutils.NodeResourceID) ([]string, error) {
 
-	asgInput := autoscaling.DetachInstancesInput{
-		AutoScalingGroupName:           asgName,
-		InstanceIds:                    instanceIDs,
+	// Grab the instanceIDs once as it is used multiple times throughout the
+	// scale in event.
+	var instanceIDs []string
+
+	// isMissing tracks the total number of instance deemed missing from the
+	// ASG to provide some user context.
+	var isMissing int
+
+	for _, node := range ids {
+
+		// found identifies whether this individual node has been located
+		// within the ASG.
+		var found bool
+
+		// Iterate the instance within the ASG, and exit if we identify a
+		// match to continue below.
+		for _, asgIDs := range asg.Instances {
+			if node.RemoteResourceID == *asgIDs.InstanceId {
+				found = true
+				break
+			}
+		}
+
+		if found {
+			instanceIDs = append(instanceIDs, node.RemoteResourceID)
+		} else {
+			isMissing++
+		}
+	}
+
+	if isMissing > 0 {
+		return nil, fmt.Errorf("%v selected nodes are not found within ASG", isMissing)
+	}
+	return instanceIDs, nil
+}
+
+// terminateInstancesInASG handles terminating all instances passed and returns
+// an object detailing the complete status of the performed action.
+func (t *TargetPlugin) terminateInstancesInASG(ctx context.Context, ids []scaleutils.NodeResourceID) instanceTerminationResult {
+
+	var status instanceTerminationResult
+
+	for _, id := range ids {
+		activityID, err := t.terminateInstance(ctx, id.RemoteResourceID)
+		if err != nil {
+			status.appendFailure(instanceFailure{instance: id, err: err})
+			continue
+		}
+		status.appendSuccess(instanceSuccess{instance: id, activityID: activityID})
+	}
+
+	return status
+}
+
+// terminateInstancesInASG terminates a single instance within an AWS
+// AutoScaling Group. It returns any error from the API, along with the
+// activity ID from the scaling event.
+func (t *TargetPlugin) terminateInstance(ctx context.Context, id string) (*string, error) {
+
+	asgInput := autoscaling.TerminateInstanceInAutoScalingGroupInput{
+		InstanceId:                     aws.String(id),
 		ShouldDecrementDesiredCapacity: aws.Bool(true),
 	}
 
-	asgResp, err := t.asg.DetachInstancesRequest(&asgInput).Send(ctx)
+	// The underlying AWS client HTTP request includes backoff and retry in the
+	// event of errors such as timeouts and rate-limiting. There is therefore
+	// no value in retrying requests that fail.
+	resp, err := t.asg.TerminateInstanceInAutoScalingGroupRequest(&asgInput).Send(ctx)
 	if err != nil {
-		return fmt.Errorf("failed to detach intances from Autoscaling Group: %v", err)
+		return nil, err
 	}
 
-	// Identify the activities that were created as a result of the detachment
-	// request so that we can go ahead and track these to completion.
-	var activityIDs []string
-
-	for _, activity := range asgResp.Activities {
-		activityIDs = append(activityIDs, *activity.ActivityId)
+	// It's unknown whether this will ever hit in the event the return error is
+	// nil, but we should protect against a nil pointer error. The ActivityId
+	// is required, therefore if Activity is not nil, this should be there.
+	if resp.Activity == nil {
+		return nil, errors.New("AWS returned nil activity response object")
 	}
-
-	// Confirm that the detachments complete before moving on. I (jrasell) am
-	// not exactly sure what happens if we terminate an instance which is still
-	// detaching from an ASG, but we might as well avoid finding out if we can.
-	err = t.ensureActivitiesComplete(ctx, activityIDs, *asgName)
-	if err != nil {
-		return fmt.Errorf("failed to detached instances from AutoScaling Group: %v", err)
-	}
-	return nil
-}
-
-func (t *TargetPlugin) terminateInstances(ctx context.Context, instanceIDs []string) error {
-
-	ec2Input := ec2.TerminateInstancesInput{InstanceIds: instanceIDs}
-
-	// TODO(jrasell) the response includes information about instance status
-	//  changes which we may want to validate in the future.
-	_, err := t.ec2.TerminateInstancesRequest(&ec2Input).Send(ctx)
-	if err != nil {
-		return fmt.Errorf("failed to terminate EC2 intances: %v", err)
-	}
-
-	// Confirm that the instances have indeed terminated properly. This allows
-	// us to handle reconciliation if the error is transient, or at least
-	// allows operators to see the error and perform manual actions to resolve.
-	err = t.ensureInstancesTerminate(ctx, instanceIDs)
-	if err != nil {
-		return fmt.Errorf("failed to terminate EC2 instances: %v", err)
-	}
-	return nil
+	return resp.Activity.ActivityId, nil
 }
 
 func (t *TargetPlugin) describeASG(ctx context.Context, asgName string) (*autoscaling.AutoScalingGroup, error) {
@@ -232,7 +295,7 @@ func (t *TargetPlugin) describeActivities(ctx context.Context, asgName string, i
 	return resp.Activities, nil
 }
 
-func (t *TargetPlugin) ensureActivitiesComplete(ctx context.Context, ids []string, asg string) error {
+func (t *TargetPlugin) ensureActivitiesComplete(ctx context.Context, asg string, ids []string) error {
 
 	f := func(ctx context.Context) (bool, error) {
 
@@ -258,40 +321,6 @@ func (t *TargetPlugin) ensureActivitiesComplete(ctx context.Context, ids []strin
 			return true, nil
 		}
 		return false, fmt.Errorf("waiting for %v activities to finish", len(ids))
-	}
-
-	return retry(ctx, defaultRetryInterval, defaultRetryLimit, f)
-}
-
-func (t *TargetPlugin) ensureInstancesTerminate(ctx context.Context, ids []string) error {
-
-	f := func(ctx context.Context) (bool, error) {
-
-		// Ensure we use IncludeAllInstances otherwise instances in a shutting
-		// down state will not be reported resulting in the function returning
-		// sooner than it should.
-		input := ec2.DescribeInstanceStatusInput{InstanceIds: ids, IncludeAllInstances: aws.Bool(true)}
-
-		resp, err := t.ec2.DescribeInstanceStatusRequest(&input).Send(ctx)
-		if err != nil {
-			return true, err
-		}
-
-		// Reset the instance IDs we want to check so this can be populated again
-		// once we have processed their current status information.
-		ids = []string{}
-
-		for _, instanceStatus := range resp.InstanceStatuses {
-			if instanceStatus.InstanceState.Name != ec2.InstanceStateNameTerminated {
-				ids = append(ids, *instanceStatus.InstanceId)
-			}
-		}
-
-		// If we dont have any remaining IDs to check, we can finish.
-		if len(ids) == 0 {
-			return true, nil
-		}
-		return false, fmt.Errorf("waiting for %v instances to terminate", len(ids))
 	}
 
 	return retry(ctx, defaultRetryInterval, defaultRetryLimit, f)

--- a/plugins/builtin/target/aws-asg/plugin/aws.go
+++ b/plugins/builtin/target/aws-asg/plugin/aws.go
@@ -137,13 +137,16 @@ func (t *TargetPlugin) scaleIn(ctx context.Context, asg *autoscaling.AutoScaling
 	result := t.terminateInstancesInASG(ctx, nodeResourceIDs)
 	result.logResults(log)
 
+	// Capture any post-termination task errors.
+	var failedTaskErr, successTaskErr error
+
 	// If we have any failures, perform our revert so we don't leave nodes in
 	// an undesired state.
 	if result.lenFailure() > 0 {
-		t.clusterUtils.RunPostScaleInTasksOnFailure(result.failedIDs())
+		failedTaskErr = t.clusterUtils.RunPostScaleInTasksOnFailure(result.failedIDs())
 	}
 
-	// If we had successfully termination from the ASG, track these activities
+	// If we had successful terminations from the ASG, track these activities
 	// until completion. A failure here should not fail the scaling activity as
 	// AWS should honour the contract, it could be a case of there being
 	// slowness in the AWS system and us timing out.
@@ -159,15 +162,22 @@ func (t *TargetPlugin) scaleIn(ctx context.Context, asg *autoscaling.AutoScaling
 		eWriter.write(ctx, scalingEventTerminate)
 
 		// Run any post scale in tasks that are desired.
-		if err := t.clusterUtils.RunPostScaleInTasks(ctx, config, result.successfulIDs()); err != nil {
-			return fmt.Errorf("failed to perform post-scale Nomad scale in tasks: %v", err)
-		}
+		successTaskErr = t.clusterUtils.RunPostScaleInTasks(ctx, config, result.successfulIDs())
 	}
 
+	// The tasks run on nodes that have been successfully terminated should not
+	// cause a failure of the scaling pipeline.
+	if successTaskErr != nil {
+		t.logger.Error("failed to perform post-scale Nomad scale in tasks", "error", successTaskErr)
+	}
+
+	// In the event of a partial failure, we want to understand whether we
+	// managed to reconcile the nodes that were not terminated before failing
+	// the pipeline.
 	if result.lenFailure() > 0 && result.lenSuccess() > 0 {
 		log.Warn("partial scaling success",
 			"success_num", result.lenSuccess(), "failed_num", result.lenFailure())
-		return nil
+		return failedTaskErr
 	}
 	return result.errorOrNil()
 }

--- a/plugins/builtin/target/aws-asg/plugin/aws_test.go
+++ b/plugins/builtin/target/aws-asg/plugin/aws_test.go
@@ -4,9 +4,71 @@ import (
 	"errors"
 	"testing"
 
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/autoscaling"
+	"github.com/hashicorp/nomad-autoscaler/sdk/helper/scaleutils"
 	"github.com/hashicorp/nomad/api"
 	"github.com/stretchr/testify/assert"
 )
+
+func Test_instancesBelongToASG(t *testing.T) {
+	testCases := []struct {
+		inputASG            *autoscaling.AutoScalingGroup
+		inputIDs            []scaleutils.NodeResourceID
+		expectedOutputList  []string
+		expectedOutputError error
+		name                string
+	}{
+		{
+			inputASG: &autoscaling.AutoScalingGroup{
+				Instances: []autoscaling.Instance{
+					{InstanceId: aws.String("i-08d2c60605d210f51")},
+					{InstanceId: aws.String("i-08d2c60605d210f52")},
+					{InstanceId: aws.String("i-08d2c60605d210f53")},
+					{InstanceId: aws.String("i-08d2c60605d210f54")},
+					{InstanceId: aws.String("i-08d2c60605d210f55")},
+				},
+			},
+			inputIDs: []scaleutils.NodeResourceID{
+				{RemoteResourceID: "i-08d2c60605d210f51"},
+				{RemoteResourceID: "i-08d2c60605d210f54"},
+			},
+			expectedOutputList: []string{
+				"i-08d2c60605d210f51",
+				"i-08d2c60605d210f54",
+			},
+			expectedOutputError: nil,
+			name:                "multiple matches with zero failure",
+		},
+		{
+			inputASG: &autoscaling.AutoScalingGroup{
+				Instances: []autoscaling.Instance{
+					{InstanceId: aws.String("i-08d2c60605d210f51")},
+					{InstanceId: aws.String("i-08d2c60605d210f52")},
+					{InstanceId: aws.String("i-08d2c60605d210f53")},
+					{InstanceId: aws.String("i-08d2c60605d210f54")},
+					{InstanceId: aws.String("i-08d2c60605d210f55")},
+				},
+			},
+			inputIDs: []scaleutils.NodeResourceID{
+				{RemoteResourceID: "i-08d2c60605d210f51"},
+				{RemoteResourceID: "i-08d2c60605d210f54"},
+				{RemoteResourceID: "i-08d2c60605d210f58"},
+			},
+			expectedOutputList:  nil,
+			expectedOutputError: errors.New("1 selected nodes are not found within ASG"),
+			name:                "multiple matches with zero failure",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			actualList, actualErr := instancesBelongToASG(tc.inputASG, tc.inputIDs)
+			assert.Equal(t, tc.expectedOutputList, actualList, tc.name)
+			assert.Equal(t, tc.expectedOutputError, actualErr, tc.name)
+		})
+	}
+}
 
 func Test_awsNodeIDMap(t *testing.T) {
 	testCases := []struct {

--- a/plugins/builtin/target/aws-asg/plugin/event.go
+++ b/plugins/builtin/target/aws-asg/plugin/event.go
@@ -16,7 +16,6 @@ type scalingEvent string
 
 const (
 	scalingEventDrain     scalingEvent = "drain"
-	scalingEventDetach    scalingEvent = "detach"
 	scalingEventTerminate scalingEvent = "terminate"
 )
 

--- a/plugins/builtin/target/aws-asg/plugin/plugin.go
+++ b/plugins/builtin/target/aws-asg/plugin/plugin.go
@@ -6,7 +6,6 @@ import (
 	"strconv"
 
 	"github.com/aws/aws-sdk-go-v2/service/autoscaling"
-	"github.com/aws/aws-sdk-go-v2/service/ec2"
 	hclog "github.com/hashicorp/go-hclog"
 	"github.com/hashicorp/nomad-autoscaler/plugins"
 	"github.com/hashicorp/nomad-autoscaler/plugins/base"
@@ -52,7 +51,6 @@ type TargetPlugin struct {
 	config map[string]string
 	logger hclog.Logger
 	asg    *autoscaling.Client
-	ec2    *ec2.Client
 
 	// clusterUtils provides general cluster scaling utilities for querying the
 	// state of nodes pools and performing scaling tasks.

--- a/plugins/builtin/target/aws-asg/plugin/status.go
+++ b/plugins/builtin/target/aws-asg/plugin/status.go
@@ -1,0 +1,129 @@
+package plugin
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	hclog "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad-autoscaler/sdk/helper/scaleutils"
+)
+
+// instanceTerminationResult tracks the state when performing a scale in action
+// and helps account for the nature of the AWS API which only accepts a single
+// instanceID per request.
+type instanceTerminationResult struct {
+	failed  []instanceFailure
+	success []instanceSuccess
+}
+
+// instanceSuccess tracks the details of an instance whose terminate API call
+// returned an error.
+type instanceFailure struct {
+	instance scaleutils.NodeResourceID
+	err      error
+}
+
+// instanceSuccess tracks the details of an instance whose terminate API call
+// completed successfully.
+type instanceSuccess struct {
+	instance   scaleutils.NodeResourceID
+	activityID *string
+}
+
+// logResults is a convenience function that logs all the currently held status
+// details to their appropriate logging level.
+func (i *instanceTerminationResult) logResults(log hclog.Logger) {
+
+	for _, success := range i.success {
+		log.Debug("successfully terminated instance in ASG",
+			"instance_id", success.instance.RemoteResourceID, "node_id", success.instance.NomadNodeID)
+	}
+
+	for _, failure := range i.failed {
+		log.Error("failed to terminate instance in ASG",
+			"instance_id", failure.instance.RemoteResourceID, "node_id", failure.instance.NomadNodeID,
+			"error", failure.err)
+	}
+}
+
+// Error satisfies the error interface, outputting all errors in a nicely
+// formatted string.
+func (i *instanceTerminationResult) Error() string {
+	if i.lenFailure() < 1 {
+		return ""
+	}
+
+	points := make([]string, i.lenFailure())
+	for i, err := range i.failed {
+		points[i] = fmt.Sprintf(
+			"failed to terminate node %s with AWS ID %s: %v",
+			err.instance.NomadNodeID, err.instance.RemoteResourceID, err.err)
+	}
+	return strings.Join(points, ", ")
+}
+
+// errorOrNil returns a new error if the result contains error entries, or nil
+// if there are not any.
+func (i *instanceTerminationResult) errorOrNil() error {
+	if i.lenFailure() < 1 {
+		return nil
+	}
+	return errors.New(i.Error())
+}
+
+// activityIDs returns a list of AWS AutoScaling activity IDs which were
+// generated as a result of terminating instances.
+func (i *instanceTerminationResult) activityIDs() []string {
+
+	if i.lenSuccess() < 1 {
+		return nil
+	}
+
+	activityIDs := make([]string, i.lenSuccess())
+	for i, id := range i.success {
+		activityIDs[i] = *id.activityID
+	}
+	return activityIDs
+}
+
+// successfulIDs returns the list of instances which were unsuccessfully
+// terminated.
+func (i *instanceTerminationResult) failedIDs() []scaleutils.NodeResourceID {
+
+	if i.lenFailure() < 1 {
+		return nil
+	}
+
+	ids := make([]scaleutils.NodeResourceID, i.lenFailure())
+	for i, inst := range i.failed {
+		ids[i] = inst.instance
+	}
+	return ids
+}
+
+// successfulIDs returns the list of instances which were successfully
+// terminated.
+func (i *instanceTerminationResult) successfulIDs() []scaleutils.NodeResourceID {
+
+	if i.lenSuccess() < 1 {
+		return nil
+	}
+
+	ids := make([]scaleutils.NodeResourceID, i.lenSuccess())
+	for i, inst := range i.success {
+		ids[i] = inst.instance
+	}
+	return ids
+}
+
+func (i *instanceTerminationResult) appendFailure(err instanceFailure) {
+	i.failed = append(i.failed, err)
+}
+
+func (i *instanceTerminationResult) appendSuccess(inf instanceSuccess) {
+	i.success = append(i.success, inf)
+}
+
+func (i *instanceTerminationResult) lenFailure() int { return len(i.failed) }
+func (i *instanceTerminationResult) lenSuccess() int { return len(i.success) }

--- a/plugins/builtin/target/aws-asg/plugin/status_test.go
+++ b/plugins/builtin/target/aws-asg/plugin/status_test.go
@@ -1,0 +1,448 @@
+package plugin
+
+import (
+	"errors"
+	"testing"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/hashicorp/nomad-autoscaler/sdk/helper/scaleutils"
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_instanceTerminationResult_Error(t *testing.T) {
+	testCases := []struct {
+		inputResult    *instanceTerminationResult
+		expectedOutput string
+		name           string
+	}{
+		{
+			inputResult:    &instanceTerminationResult{},
+			expectedOutput: "",
+			name:           "empty input result",
+		},
+		{
+			inputResult: &instanceTerminationResult{
+				failed: []instanceFailure{
+					{
+						instance: scaleutils.NodeResourceID{
+							NomadNodeID:      "711eb2aa-48cc-2dc7-32fa-b359878121cd",
+							RemoteResourceID: "i-08d2c60605d210f57",
+						},
+						err: errors.New("this is the error you're looking for"),
+					},
+				},
+			},
+			expectedOutput: "failed to terminate node 711eb2aa-48cc-2dc7-32fa-b359878121cd with AWS ID i-08d2c60605d210f57: this is the error you're looking for",
+			name:           "single input result error",
+		},
+		{
+			inputResult: &instanceTerminationResult{
+				failed: []instanceFailure{
+					{
+						instance: scaleutils.NodeResourceID{
+							NomadNodeID:      "711eb2aa-48cc-2dc7-32fa-b359878121cd",
+							RemoteResourceID: "i-08d2c60605d210f57",
+						},
+						err: errors.New("this is the error you're looking for"),
+					},
+					{
+						instance: scaleutils.NodeResourceID{
+							NomadNodeID:      "711eb2aa-48cc-2dc7-32fa-b359878121ce",
+							RemoteResourceID: "i-08d2c60605d210f58",
+						},
+						err: errors.New("this isn't the error you're looking for"),
+					},
+				},
+			},
+			expectedOutput: "failed to terminate node 711eb2aa-48cc-2dc7-32fa-b359878121cd with AWS ID i-08d2c60605d210f57: this is the error you're looking for, failed to terminate node 711eb2aa-48cc-2dc7-32fa-b359878121ce with AWS ID i-08d2c60605d210f58: this isn't the error you're looking for",
+			name:           "multiple input result error",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			assert.Equal(t, tc.expectedOutput, tc.inputResult.Error(), tc.name)
+		})
+	}
+}
+
+func Test_instanceTerminationResult_errorOrNil(t *testing.T) {
+	testCases := []struct {
+		inputResult    *instanceTerminationResult
+		expectedOutput error
+		name           string
+	}{
+		{
+			inputResult:    &instanceTerminationResult{},
+			expectedOutput: nil,
+			name:           "no errors",
+		},
+		{
+			inputResult: &instanceTerminationResult{
+				failed: []instanceFailure{
+					{
+						instance: scaleutils.NodeResourceID{
+							NomadNodeID:      "711eb2aa-48cc-2dc7-32fa-b359878121cd",
+							RemoteResourceID: "i-08d2c60605d210f57",
+						},
+						err: errors.New("this is the error you're looking for"),
+					},
+				},
+			},
+			expectedOutput: errors.New("failed to terminate node 711eb2aa-48cc-2dc7-32fa-b359878121cd with AWS ID i-08d2c60605d210f57: this is the error you're looking for"),
+			name:           "error",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			assert.Equal(t, tc.expectedOutput, tc.inputResult.errorOrNil(), tc.name)
+		})
+	}
+}
+
+func Test_instanceTerminationResult_activityIDs(t *testing.T) {
+	testCases := []struct {
+		inputResult    *instanceTerminationResult
+		expectedOutput []string
+		name           string
+	}{
+		{
+			inputResult:    &instanceTerminationResult{},
+			expectedOutput: nil,
+			name:           "empty result",
+		},
+		{
+			inputResult: &instanceTerminationResult{
+				success: []instanceSuccess{
+					{
+						instance: scaleutils.NodeResourceID{
+							NomadNodeID:      "711eb2aa-48cc-2dc7-32fa-b359878121cd",
+							RemoteResourceID: "i-08d2c60605d210f57",
+						},
+						activityID: aws.String("f9f2d65b-f1f2-43e7-b46d-d86756459690"),
+					},
+					{
+						instance: scaleutils.NodeResourceID{
+							NomadNodeID:      "711eb2aa-48cc-2dc7-32fa-b359878121cf",
+							RemoteResourceID: "i-08d2c60605d210f59",
+						},
+						activityID: aws.String("f9f2d65b-f1f2-43e7-b46d-d86756459691"),
+					},
+				},
+				failed: []instanceFailure{
+					{
+						instance: scaleutils.NodeResourceID{
+							NomadNodeID:      "711eb2aa-48cc-2dc7-32fa-b359878121ce",
+							RemoteResourceID: "i-08d2c60605d210f58",
+						},
+						err: errors.New("this is the error you're looking for"),
+					},
+				},
+			},
+			expectedOutput: []string{
+				"f9f2d65b-f1f2-43e7-b46d-d86756459690",
+				"f9f2d65b-f1f2-43e7-b46d-d86756459691",
+			},
+			name: "failed and success within result",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			assert.Equal(t, tc.expectedOutput, tc.inputResult.activityIDs(), tc.name)
+		})
+	}
+}
+
+func Test_instanceTerminationResult_failedIDs(t *testing.T) {
+	testCases := []struct {
+		inputResult    *instanceTerminationResult
+		expectedOutput []scaleutils.NodeResourceID
+		name           string
+	}{
+		{
+			inputResult:    &instanceTerminationResult{},
+			expectedOutput: nil,
+			name:           "empty input result",
+		},
+		{
+			inputResult: &instanceTerminationResult{
+				failed: []instanceFailure{
+					{
+						instance: scaleutils.NodeResourceID{
+							NomadNodeID:      "711eb2aa-48cc-2dc7-32fa-b359878121cd",
+							RemoteResourceID: "i-08d2c60605d210f57",
+						},
+						err: errors.New("this is the error you're looking for"),
+					},
+				},
+			},
+			expectedOutput: []scaleutils.NodeResourceID{
+				{
+					NomadNodeID:      "711eb2aa-48cc-2dc7-32fa-b359878121cd",
+					RemoteResourceID: "i-08d2c60605d210f57",
+				},
+			},
+			name: "single entry",
+		},
+		{
+			inputResult: &instanceTerminationResult{
+				failed: []instanceFailure{
+					{
+						instance: scaleutils.NodeResourceID{
+							NomadNodeID:      "711eb2aa-48cc-2dc7-32fa-b359878121cd",
+							RemoteResourceID: "i-08d2c60605d210f57",
+						},
+						err: errors.New("this is the error you're looking for"),
+					},
+					{
+						instance: scaleutils.NodeResourceID{
+							NomadNodeID:      "711eb2aa-48cc-2dc7-32fa-b359878121ce",
+							RemoteResourceID: "i-08d2c60605d210f58",
+						},
+						err: errors.New("this isn't the error you're looking for"),
+					},
+				},
+			},
+			expectedOutput: []scaleutils.NodeResourceID{
+				{
+					NomadNodeID:      "711eb2aa-48cc-2dc7-32fa-b359878121cd",
+					RemoteResourceID: "i-08d2c60605d210f57",
+				},
+				{
+					NomadNodeID:      "711eb2aa-48cc-2dc7-32fa-b359878121ce",
+					RemoteResourceID: "i-08d2c60605d210f58",
+				},
+			},
+			name: "multiple entries",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			assert.Equal(t, tc.expectedOutput, tc.inputResult.failedIDs(), tc.name)
+		})
+	}
+}
+
+func Test_instanceTerminationResult_successfulIDs(t *testing.T) {
+	testCases := []struct {
+		inputResult    *instanceTerminationResult
+		expectedOutput []scaleutils.NodeResourceID
+		name           string
+	}{
+		{
+			inputResult:    &instanceTerminationResult{},
+			expectedOutput: nil,
+			name:           "empty input result",
+		},
+		{
+			inputResult: &instanceTerminationResult{
+				success: []instanceSuccess{
+					{
+						instance: scaleutils.NodeResourceID{
+							NomadNodeID:      "711eb2aa-48cc-2dc7-32fa-b359878121cd",
+							RemoteResourceID: "i-08d2c60605d210f57",
+						},
+						activityID: aws.String("f9f2d65b-f1f2-43e7-b46d-d86756459690"),
+					},
+				},
+			},
+			expectedOutput: []scaleutils.NodeResourceID{
+				{
+					NomadNodeID:      "711eb2aa-48cc-2dc7-32fa-b359878121cd",
+					RemoteResourceID: "i-08d2c60605d210f57",
+				},
+			},
+			name: "single entry",
+		},
+		{
+			inputResult: &instanceTerminationResult{
+				success: []instanceSuccess{
+					{
+						instance: scaleutils.NodeResourceID{
+							NomadNodeID:      "711eb2aa-48cc-2dc7-32fa-b359878121cd",
+							RemoteResourceID: "i-08d2c60605d210f57",
+						},
+						activityID: aws.String("f9f2d65b-f1f2-43e7-b46d-d86756459690"),
+					},
+					{
+						instance: scaleutils.NodeResourceID{
+							NomadNodeID:      "711eb2aa-48cc-2dc7-32fa-b359878121ce",
+							RemoteResourceID: "i-08d2c60605d210f58",
+						},
+						activityID: aws.String("f9f2d65b-f1f2-43e7-b46d-d86756459691"),
+					},
+				},
+			},
+			expectedOutput: []scaleutils.NodeResourceID{
+				{
+					NomadNodeID:      "711eb2aa-48cc-2dc7-32fa-b359878121cd",
+					RemoteResourceID: "i-08d2c60605d210f57",
+				},
+				{
+					NomadNodeID:      "711eb2aa-48cc-2dc7-32fa-b359878121ce",
+					RemoteResourceID: "i-08d2c60605d210f58",
+				},
+			},
+			name: "multiple entries",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			assert.Equal(t, tc.expectedOutput, tc.inputResult.successfulIDs(), tc.name)
+		})
+	}
+}
+
+func Test_instanceTerminationResult_appendFailure(t *testing.T) {
+	testCases := []struct {
+		inputResult *instanceTerminationResult
+		inputErr    instanceFailure
+		name        string
+	}{
+		{
+			inputResult: &instanceTerminationResult{},
+			inputErr: instanceFailure{
+				instance: scaleutils.NodeResourceID{
+					NomadNodeID:      "711eb2aa-48cc-2dc7-32fa-b359878121cb",
+					RemoteResourceID: "i-08d2c60605d210f56",
+				},
+				err: errors.New("this isn't the error you're looking for"),
+			},
+			name: "empty input result",
+		},
+		{
+			inputResult: &instanceTerminationResult{
+				failed: []instanceFailure{
+					{
+						instance: scaleutils.NodeResourceID{
+							NomadNodeID:      "711eb2aa-48cc-2dc7-32fa-b359878121cd",
+							RemoteResourceID: "i-08d2c60605d210f57",
+						},
+						err: errors.New("this is the error you're looking for"),
+					},
+				},
+			},
+			inputErr: instanceFailure{
+				instance: scaleutils.NodeResourceID{
+					NomadNodeID:      "711eb2aa-48cc-2dc7-32fa-b359878121cb",
+					RemoteResourceID: "i-08d2c60605d210f56",
+				},
+				err: errors.New("this isn't the error you're looking for"),
+			},
+			name: "non-empty input result",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			tc.inputResult.appendFailure(tc.inputErr)
+			assert.Contains(t, tc.inputResult.failed, tc.inputErr, tc.name)
+		})
+	}
+}
+
+func Test_instanceTerminationResult_appendSuccess(t *testing.T) {
+	testCases := []struct {
+		inputResult *instanceTerminationResult
+		inputInf    instanceSuccess
+		name        string
+	}{
+		{
+			inputResult: &instanceTerminationResult{},
+			inputInf: instanceSuccess{
+				instance: scaleutils.NodeResourceID{
+					NomadNodeID:      "711eb2aa-48cc-2dc7-32fa-b359878121cb",
+					RemoteResourceID: "i-08d2c60605d210f56",
+				},
+				activityID: aws.String("f9f2d65b-f1f2-43e7-b46d-d86756459699"),
+			},
+			name: "empty input result",
+		},
+		{
+			inputResult: &instanceTerminationResult{
+				success: []instanceSuccess{
+					{
+						instance: scaleutils.NodeResourceID{
+							NomadNodeID:      "711eb2aa-48cc-2dc7-32fa-b359878121cd",
+							RemoteResourceID: "i-08d2c60605d210f57",
+						},
+						activityID: aws.String("f9f2d65b-f1f2-43e7-b46d-d86756459690"),
+					},
+				},
+			},
+			inputInf: instanceSuccess{
+				instance: scaleutils.NodeResourceID{
+					NomadNodeID:      "711eb2aa-48cc-2dc7-32fa-b359878121cb",
+					RemoteResourceID: "i-08d2c60605d210f56",
+				},
+				activityID: aws.String("f9f2d65b-f1f2-43e7-b46d-d86756459699"),
+			},
+			name: "non-empty input result",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			tc.inputResult.appendSuccess(tc.inputInf)
+			assert.Contains(t, tc.inputResult.success, tc.inputInf, tc.name)
+		})
+	}
+}
+
+func Test_instanceTerminationResult_lenFailure(t *testing.T) {
+	testCases := []struct {
+		inputResult    *instanceTerminationResult
+		expectedOutput int
+		name           string
+	}{
+		{
+			inputResult:    &instanceTerminationResult{},
+			expectedOutput: 0,
+			name:           "empty input result",
+		},
+		{
+			inputResult: &instanceTerminationResult{
+				failed: make([]instanceFailure, 13),
+			},
+			expectedOutput: 13,
+			name:           "non-zero input result",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			assert.Equal(t, tc.expectedOutput, tc.inputResult.lenFailure(), tc.name)
+		})
+	}
+}
+
+func Test_instanceTerminationResult_lenSuccess(t *testing.T) {
+	testCases := []struct {
+		inputResult    *instanceTerminationResult
+		expectedOutput int
+		name           string
+	}{
+		{
+			inputResult:    &instanceTerminationResult{},
+			expectedOutput: 0,
+			name:           "empty input result",
+		},
+		{
+			inputResult: &instanceTerminationResult{
+				success: make([]instanceSuccess, 13),
+			},
+			expectedOutput: 13,
+			name:           "non-zero input result",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			assert.Equal(t, tc.expectedOutput, tc.inputResult.lenSuccess(), tc.name)
+		})
+	}
+}

--- a/sdk/helper/scaleutils/cluster.go
+++ b/sdk/helper/scaleutils/cluster.go
@@ -64,7 +64,7 @@ func (c *ClusterScaleUtils) RunPreScaleInTasks(ctx context.Context, cfg map[stri
 		return nil, errors.New("required ClusterNodeIDLookupFunc not set")
 	}
 
-	nodes, err := c.identifyScaleInNodes(cfg, num)
+	nodes, err := c.IdentifyScaleInNodes(cfg, num)
 	if err != nil {
 		return nil, err
 	}
@@ -72,7 +72,7 @@ func (c *ClusterScaleUtils) RunPreScaleInTasks(ctx context.Context, cfg map[stri
 	// Technically we do not need this information until after the nodes have
 	// been drained. However, this doesn't change cluster state and so do this
 	// first to make sure there are no issues in translating.
-	nodeResourceIDs, err := c.identifyScaleInRemoteIDs(nodes)
+	nodeResourceIDs, err := c.IdentifyScaleInRemoteIDs(nodes)
 	if err != nil {
 		return nil, err
 	}
@@ -80,7 +80,7 @@ func (c *ClusterScaleUtils) RunPreScaleInTasks(ctx context.Context, cfg map[stri
 	// Drain the nodes.
 	// TODO(jrasell) we should try some reconciliation here, where we identify
 	//  failed nodes and continue with nodes that drained successfully.
-	if err := c.drainNodes(ctx, cfg, nodeResourceIDs); err != nil {
+	if err := c.DrainNodes(ctx, cfg, nodeResourceIDs); err != nil {
 		return nil, err
 	}
 	c.log.Info("pre scale-in tasks now complete")
@@ -88,7 +88,7 @@ func (c *ClusterScaleUtils) RunPreScaleInTasks(ctx context.Context, cfg map[stri
 	return nodeResourceIDs, nil
 }
 
-func (c *ClusterScaleUtils) identifyScaleInNodes(cfg map[string]string, num int) ([]*api.NodeListStub, error) {
+func (c *ClusterScaleUtils) IdentifyScaleInNodes(cfg map[string]string, num int) ([]*api.NodeListStub, error) {
 
 	// The Nomad Autoscaler can only handle node class identifiers currently
 	// and therefore we just set that up. In the future if we wish to expand on
@@ -152,7 +152,7 @@ func (c *ClusterScaleUtils) identifyScaleInNodes(cfg map[string]string, num int)
 	return out, nil
 }
 
-func (c *ClusterScaleUtils) identifyScaleInRemoteIDs(nodes []*api.NodeListStub) ([]NodeResourceID, error) {
+func (c *ClusterScaleUtils) IdentifyScaleInRemoteIDs(nodes []*api.NodeListStub) ([]NodeResourceID, error) {
 
 	var (
 		out  []NodeResourceID

--- a/sdk/helper/scaleutils/node_drain.go
+++ b/sdk/helper/scaleutils/node_drain.go
@@ -156,14 +156,20 @@ func (c *ClusterScaleUtils) monitorNodeDrain(ctx context.Context, nodeID string,
 // were unable to be successfully terminated. The current tasks are:
 //
 //   - modify node eligibility to true
-func (c *ClusterScaleUtils) RunPostScaleInTasksOnFailure(nodes []NodeResourceID) {
+func (c *ClusterScaleUtils) RunPostScaleInTasksOnFailure(nodes []NodeResourceID) error {
+
+	var mErr *multierror.Error
+
 	for _, node := range nodes {
 		resp, err := c.client.Nodes().ToggleEligibility(node.NomadNodeID, true, nil)
 		if err != nil {
-			c.log.Error("failed to toggle node eligibility", "error", err)
+			mErr = multierror.Append(mErr, fmt.Errorf("failed to toggle eligibility on node %s: %v",
+				node.NomadNodeID, err))
 			continue
 		}
 		c.log.Info("successfully toggled node eligibility",
 			"node_id", node.NomadNodeID, "evals", resp.EvalIDs)
 	}
+
+	return formattedMultiError(mErr)
 }

--- a/sdk/helper/scaleutils/node_drain.go
+++ b/sdk/helper/scaleutils/node_drain.go
@@ -17,10 +17,10 @@ const (
 	defaultNodeIgnoreSystemJobs = false
 )
 
-// drainNodes iterates the provided nodeID list and performs a drain on each
+// DrainNodes iterates the provided nodeID list and performs a drain on each
 // one. Each node drain is monitored and events logged until the context is
 // closed or all drains reach a terminal state.
-func (c *ClusterScaleUtils) drainNodes(ctx context.Context, cfg map[string]string, nodes []NodeResourceID) error {
+func (c *ClusterScaleUtils) DrainNodes(ctx context.Context, cfg map[string]string, nodes []NodeResourceID) error {
 
 	drainSpec, err := drainSpec(cfg)
 	if err != nil {
@@ -150,4 +150,20 @@ func (c *ClusterScaleUtils) monitorNodeDrain(ctx context.Context, nodeID string,
 		}
 	}
 	return ctx.Err()
+}
+
+// RunPostScaleInTasksOnFailure is a utility that runs tasks on nodes which
+// were unable to be successfully terminated. The current tasks are:
+//
+//   - modify node eligibility to true
+func (c *ClusterScaleUtils) RunPostScaleInTasksOnFailure(nodes []NodeResourceID) {
+	for _, node := range nodes {
+		resp, err := c.client.Nodes().ToggleEligibility(node.NomadNodeID, true, nil)
+		if err != nil {
+			c.log.Error("failed to toggle node eligibility", "error", err)
+			continue
+		}
+		c.log.Info("successfully toggled node eligibility",
+			"node_id", node.NomadNodeID, "evals", resp.EvalIDs)
+	}
 }

